### ISSUE TITLE
Plugins not working under ubuntu 22.04

### ISSUE THREAD

**Issue by LennartG-Sve** [jongough/testplugin_pi#299]
I've tested all available plugins under xubuntu 22.04 and got a rather disturbing result. The system reports, using 'cat /etc/os-release', as:
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian

And this is the outcome:

1/ OpenCPN with standard plugins (GRIB, Chart Downloader, Dashboard 1.2 and WMM 1.1), already installed, ok.
2/ o-charts 1.0.34 already installed, ok
3/ From master catalogue: Celestial navigation, Climatology, Dashboard Tactics, Findit, GPS Odometer, Javascript, NMEA Converter, OCPN Draw, Plots, Polar, Pypilot, S63, Testplugin (230):, Watchdog and Weatherfax (all available)
4/ From beta catalogue: Dashboard Tactics, OCPN Draw, Tactics (all available)

All of the plugins listed under 3/ and 4/ gives the same result: 
Installs ok but message is then 'Plugin is not compatible with this version of OpenCPN, and will be uninstalled'.

5/ Checking opencpn.log shows that all of the failed plugins share the same error message:
Plugin depends on library "libwx-baseu-3.0.so.0", but the main module was built for "libwx-baseu-3.2.so.0"

Appears we have a wx-version problem.
My xubuntu 22.04 system is a standard installation, fully updated using the standard procedures for linux. It also has both wx3.0 and wx3.2 fully and properly installed, as far as I can tell.

**Comment by jongough** [comment]
Check in advanced plugin options what is set for wxwidgets. Is it 3.2 or 3.0? It sounds like it is installing the 3.0 version then finding it is not compatible.
The installation process in ocpn should show you what is being installed.

**Comment by rgleason** [comment]
Lennart, also in addition to Jon's suggestion it might be helpful to know if something changed between TP 1.0.226 and 1.0.230. Can you import just one or two TP1.0.226 plugind to see if they fail?

**Comment by LennartG-Sve** [comment]
Jon, please explain 'advanced plugin options'. I'm using 5.7.1 as downloaded. not compiling OpenCPN myself so how can I tell? The OpenCPN log file however states 'wxWidgets 3.2.1 Linux 64 bit wxGTK'

rgleason: how am I to tell the TP version used for the plugins? Also note that all plugins available failed with the same error message. None but the built-in plus o-charts worked.

**Comment by rgleason** [comment]
Sorry, under commits for a given plugin  https://github.com/rgleason/weatherfax_pi/commits/master
you will see  

[fix manual wxfax icon](https://github.com/rgleason/weatherfax_pi/commit/6a0d1314cd6863032d29a48f3890e2a4d4787677)
@[rgleason](https://github.com/rgleason/weatherfax_pi/commits?author=rgleason)
rgleason committed Feb 27, 2023

[1.9.55.0 + TP 1.0.226](https://github.com/rgleason/weatherfax_pi/commit/f58e68c25c058f1dbbb695bf4a658ee107b061c0)

Now find 1.9.55 in cloudsmith   weatherfax-prod repos   
in the search bar at the top enter 1.9.55.0 + (your OS)     find the tarball.
https://cloudsmith.io/~opencpn/repos/weatherfax-prod/packages/?q=1.9.55.0

Then download the file  (hoover at icon in far right)    and use PIM  "Manual Load Plugin"

**Comment by jongough** [comment]
You need to set CatalogExpert=1 in the OpenVPN.conf/ini file in the [plugins] section before starting OpenVPN. You will then find a new setting in the plugins tab allowing you to adjust the 'platform' you are using as well as access to the alpha plugins.

**Comment by rgleason** [comment]
Good point. Jon.

**Comment by LennartG-Sve** [comment]
OK, followed what Jon suggested and changed the flavour to 'ubuntu-wx32-x86:22.04', that appears to be the closest relative.
Now all plugins loaded from both master and beta with one (two?) exception(s), ShipDriver failed to load, both the master and beta versions. No error message was given in the OpenCPN.log. 

Interesting as this, as I understand it, should be the default flavour.

**Comment by jongough** [comment]
Glad it worked.
Can you please raise an issue on opencpn about this (there 'may' be an issue already on this, I cannot remember) so that it is documented.
Thanks.

**Comment by LennartG-Sve** [comment]
Added as issue #3130

@Rasbats: The only plugins, from master/beta catalogs that loaded correctly when I started this was the built-in plus o-charts.
My only concern is that all plugins should use the same default flavor. That is also why I opened the issue. 
Please keep an eye on opencpn #3130 so that all select the same flavor.

Ready to close this one of no-one objects.

**Comment by Rasbats** [comment]
@LennartG-Sve ... ShipDriver is now in the master catalog. Runs fine on a new Ubuntu Jammy installation.

**Comment by LennartG-Sve** [comment]
@Rasbats: Nice job, now it loads just fine. One (many) step(s) ahead of me .....

**Comment by jongough** [comment]
1/ The gilename is not used in determining compactability, it is the content of the XML that does that. As 22.04 is supposed to be wx32 it is not needed in the XML. This one of the confusing things in plugin manager.

There seems to a problem if opecpn is upgraded in that the old plugin manager Catalog definition, wx30/wx32 is used which may be incorrect.

To help identify what is wrong including the full version of the plugin. I.e. 1.2.3.4, and the repository, alpha/beta/prod, along with an extract if the opencpn log showing the plugin manager installation and the failure messages would help.

**Comment by LennartG-Sve** [comment]
Jon, thanks for the info, that clears it up a lot, so the flavour should then be 'ubuntu-x86_64;22.04'. 
The opencpn issue 'Plugins on Ubuntu 22.04 (5.7.1) using wrong flavor(?)
#3130' was around the same subject and @leamas closed it yesterday as we thought it solved. Now I'm no longer sure it was but I can't reopen the issue.
However, on a second system with 5.7.1, that I installed about a month ago and have not touched since, it says 'ubuntu-x32-x86_64;22.04'. There has apparently been some confusion regarding the settings.
Funny thing that the 'wx32' version is a selectable flavour if it should not be used, or even exist.

It is also apparent that there are still some details that need be sorted out with the configuration on ubuntu 22.04 so I will back off  until you are back from you well earned vacation.

**Comment by LennartG-Sve** [comment]
Jon, sorry I missed the part about the error message in the log file. It says: 
Plugin depends on library "libwx-baseu-3.0.so.0", but the main module was built for "libwx-baseu-3.2.so.0"

That is what caused all the confusion in the first place.

**Comment by leamas** [comment]
Actually, 22.04 is the only Ubuntu version which uses wx32. This is because this is the only ubuntu version where we have both wx3.0 and wx3.2 plugins available. 

The kind of plugins loaded depends on the OpenCPN version. versions <= 5.6 loads *ubuntu-x86_64* using wx3.0.  Versions > 5.6 (basically upcoming 5.8 and related beta/RC) loads *ubuntu-wx32-x86_64*.

In the end, we are actually using Debian plugins rather than Ubuntu.  That is. on versions <= 5.6  we load *debian-11-x86_64:11*. On versions > 5.6 we  use *debian-wx32-x86_64:11*. This is handled transparently, the official ABI declared and visible in *Plugins* | *Settings* is ubuntu.

I have used x86_64 here. However, it works the same way for armhf and arm64.

**Comment by LennartG-Sve** [comment]
@leamas : Getting clearer and clearer although that raises a question: How come a plugin installed on 5.7.1 could depend on library "libwx-baseu-3.0.so.0", as noted in my initial issue description above?

**Comment by leamas** [comment]
If so, and if nothing has been done usiing Plugins |Settings, this is a bug. Most likely on the plugin side, but main opencpn before RC2 was also affected by a bug in this area.

**Comment by LennartG-Sve** [comment]
@leamas , now this is getting really interesting. I repeated the initial test on the available plugins and now, instead of giving an error message and an error in the log, OpenCPN crashes within a second, the log carries no useful information..

This happens right after the message that the plugin has been successfully installed and you clicked ok. 
Tested on, from the master catalogue: 
Celestial navigation, Climatology, Dashboard Tactics, Findit and GPS Odometer. Stopped there as it felt a common pattern.

**Comment by leamas** [comment]
Have you updated to RC2? If not, this should be the next step. Also, remove all existing plugins in ~/.local/lib/opencpn.

After that, install a single plugin. If it fails, please re-run with --loglevel debug and submit the logfile

**Comment by LennartG-Sve** [comment]
@leamas , 
1/ yes I've updated to the latest RC2 this morning. My current version is 'OpenCPN  5.7.1+7936+fccc1422c-ubuntu22-04.1'
2/ The only plugin in ~/.local/lib/opencpn war libo-charts_pi.so, that has now been removed.
3/ Installed 'Celestial navigation' again, crash. Cleared the system again.
4 Ran the system with --loglevel debug and installed 'Celestial navigation' again. This is what I got:

lennart@Mysak:~$ opencpn --loglevel debug 
ALSA lib pcm_dsnoop.c:601:(snd_pcm_dsnoop_open) unable to open slave
ALSA lib pcm_dmix.c:1032:(snd_pcm_dmix_open) unable to open slave
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib pcm_oss.c:397:(_snd_pcm_oss_open) Cannot open device /dev/dsp
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib confmisc.c:160:(snd_config_get_card) Invalid field card
ALSA lib pcm_usb_stream.c:482:(_snd_pcm_usb_stream_open) Invalid card 'card'
ALSA lib pcm_dmix.c:1032:(snd_pcm_dmix_open) unable to open slave
14:23:12: Debug: Device: 0: HDA Intel PCH: ALC892 Digital (hw:0,1)
14:23:12: Debug: Device: 1: HDA Intel PCH: ALC892 Alt Analog (hw:0,2)
14:23:12: Debug: Device: 2: HDA Intel PCH: HDMI 0 (hw:0,3)
14:23:12: Debug: Device: 3: HDA Intel PCH: HDMI 1 (hw:0,7)
14:23:12: Debug: Device: 4: iec958
14:23:12: Debug: Device: 5: spdif
14:23:12: Debug: Device: 6: hdmi
14:23:12: Debug: Device: 7: jack
14:23:12: Debug: Device: 8: default
14:23:12: Debug: Device: 9: system
14:23:12: Debug: Device: 0: HDA Intel PCH: ALC892 Digital (hw:0,1)
14:23:12: Debug: Device: 1: HDA Intel PCH: ALC892 Alt Analog (hw:0,2)
14:23:12: Debug: Device: 2: HDA Intel PCH: HDMI 0 (hw:0,3)
14:23:12: Debug: Device: 3: HDA Intel PCH: HDMI 1 (hw:0,7)
14:23:12: Debug: Device: 4: iec958
14:23:12: Debug: Device: 5: spdif
14:23:12: Debug: Device: 6: hdmi
14:23:12: Debug: Device: 7: jack
14:23:12: Debug: Device: 8: default
14:23:12: Debug: Device: 9: system
14:23:12: Debug: Device: 0: HDA Intel PCH: ALC892 Digital (hw:0,1)
14:23:12: Debug: Device: 1: HDA Intel PCH: ALC892 Alt Analog (hw:0,2)
14:23:12: Debug: Device: 2: HDA Intel PCH: HDMI 0 (hw:0,3)
14:23:12: Debug: Device: 3: HDA Intel PCH: HDMI 1 (hw:0,7)
14:23:12: Debug: Device: 4: iec958
14:23:12: Debug: Device: 5: spdif
14:23:12: Debug: Device: 6: hdmi
14:23:12: Debug: Device: 7: jack
14:23:12: Debug: Device: 8: default
14:23:12: Debug: Device: 9: system
14:23:12: Debug: Device: 0: HDA Intel PCH: ALC892 Digital (hw:0,1)
14:23:12: Debug: Device: 1: HDA Intel PCH: ALC892 Alt Analog (hw:0,2)
14:23:12: Debug: Device: 2: HDA Intel PCH: HDMI 0 (hw:0,3)
14:23:12: Debug: Device: 3: HDA Intel PCH: HDMI 1 (hw:0,7)
14:23:12: Debug: Device: 4: iec958
14:23:12: Debug: Device: 5: spdif
14:23:12: Debug: Device: 6: hdmi
14:23:12: Debug: Device: 7: jack
14:23:12: Debug: Device: 8: default
14:23:12: Debug: Device: 9: system
GLEW init success!nGenerating RSA key...
Generating x509 certificate...
Writing key and certificate to disk...
Success!
Local IPv4 address: 10.10.4.10
Opened 2 sockets for mDNS service
Service mDNS: opencpn-object-control-service.:8000
Hostname: Mysak
Sending announce
SSKTAP: 

(opencpn:2083): Gtk-CRITICAL **: 14:23:23.514: gtk_window_resize: assertion 'height > 0' failed

(opencpn:2083): Gtk-CRITICAL **: 14:23:25.557: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(opencpn:2083): Gtk-CRITICAL **: 14:23:25.557: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar
SSKTAP: 
02:23:36 PM: Warning: Unknown image data format.
../src/gtk/bitmap.cpp(581): assert "image.IsOk()" failed in wxBitmap(): invalid image
02:23:36 PM: Loading toolbar icon: /home/lennart/.local/share/opencpn/plugins/celestial_navigation_pi/data/celestial_navigation.svg
02:23:36 PM: Loading toolbar icon: /home/lennart/.local/share/opencpn/plugins/celestial_navigation_pi/data/celestial_navigation_rollover.svg
02:23:36 PM: Loading toolbar icon: /home/lennart/.local/share/opencpn/plugins/celestial_navigation_pi/data/celestial_navigation_toggled.svg
02:23:36 PM: Debug: Using icon path: /home/lennart/.local/share/opencpn/plugins/celestial_navigation_pi/data/celestial_navigation_panel.png
../src/common/image.cpp(1876): assert "IsOk()" failed in GetData(): invalid image
Segmenteringsfel (minnesutskrift skapad)
lennart@Mysak:~$ 

The last text in Swedish states 'Segmentation fault (crash log generated)', have not yet looked for the crash log.

**Comment by LennartG-Sve** [comment]
@leamas , found the crash log, it is just above 30 MB, probably to big to attach even if zipped. Any other option?

**Comment by leamas** [comment]
Should not be a problem to attach if compressed If it is, remove it and re-run, this log will be much smaller

**Comment by LennartG-Sve** [comment]
[_usr_bin_opencpn.1000.crash.zip](https://github.com/jongough/testplugin_pi/files/11178992/_usr_bin_opencpn.1000.crash.zip)

**Comment by LennartG-Sve** [comment]
Opencpn debug log attached.
[opencpn-debug.log.zip](https://github.com/jongough/testplugin_pi/files/11179873/opencpn-debug.log.zip)

**Comment by bdbcat** [comment]
I examined the log file.
It looks pretty clear to me that the linux flavor you are using is incorrect for O571 on Ubuntu 22.04
The interesting log entry are these:

```
14:23:17.189 MESSAGE ocpn_app.cpp:1229 wxWidgets version: wxWidgets 3.2.1 Linux 64 bit wxGTK
.
.
.
14:23:36.328   DEBUG plugin_handler.cpp:178 Plugin: setting up, name: Celestial Navigation
14:23:36.328   DEBUG plugin_handler.cpp:179 Plugin: init: abi: ubuntu-x86_64, abi_version: 22.04, major ver: 22
14:23:36.328   DEBUG plugin_handler.cpp:201 Host: init: abi: ubuntu-x86_64, abi_version: 22.04, major ver: 22
14:23:36.328   DEBUG plugin_handler.cpp:321 Found matching abi version 22.04
14:23:36.328   DEBUG plugin_handler.cpp:326 Plugin compatibility check Final: ACCEPTED: Celestial Navigation

```
This is the error point.  He loads a plugin that is not built for wx32, because probably the Host ABI description ( ubuntu-x86_64) is incorrect.

I attach snap of plugin manager setting that should be used for O571+ on Ubuntu 22.04
![pin22 04](https://user-images.githubusercontent.com/1015682/230669505-6525be4b-c130-4347-b541-26acebb8c46f.jpeg)

**Comment by bdbcat** [comment]
Just to be clear here:
Are you building OCPN 5.7.1. from source, or installing from Beta PPA?

**Comment by leamas** [comment]
Adding to this https://github.com/OpenCPN/OpenCPN/issues/3130, where it actually looks like the ABI setting is as it should by default. 

In other words, if you use the the Catalog | Settings to set the Active Setting to _Default setting_ it should revert to _ubuntu-wx32-x86_64:22.04_. Does it?

**Comment by LennartG-Sve** [comment]
@bdbcat , I'm using the PPA version, downloaded yesterday,  version is 'OpenCPN 5.7.1+7936+fccc1422c-ubuntu22-04.1'

@leamas , at boot it initiates as 'ubuntu-wx32-x86_64:22.04' and the select button states 'Select new flavour'. Selecting 'Default' refreshes the setting to the same  'ubuntu-wx32-x86_64:22.04'.
Manually selecting 'ubuntu--x86_64:22.04' and then selecting 'Default' reverts back to  'ubuntu-wx32-x86_64:22.04'.

**Comment by leamas** [comment]
OK, that should settle it. The error you report comes from using the wrong ABI setting.  

It should be noted the Plugins | Settings dialog by default is hidden for end users for good reasons.

If you want to get something working you need to  reset the settings so it becomes _ubuntu-wx32-x86_64:22.04_. You also need to completely clear the _~/.local/lib/opencpn_ directory which now probably contains unloadable plugins. You should then be able to list and install available plugins.

**Comment by leamas** [comment]
... but then again, just to be really, really sure since we are about to release. Could you please:

  - Remove all installed plugins using `rm -f ~/.local/lib/opencpn/*`
  - Start opencpn using `--loglevel debug` 
  - Revert Active Setting to _ubuntu-wx32-x86_64:22.04_
  - Install a single plugin
  - Exit opencpn
  - Submit _~/.opencpn/opencpn.log_

**Comment by LennartG-Sve** [comment]
@leamas , followed your directions as listed.
1/ Remove ~/.opencpn/opencpn.log. Done
2/ rm -f ~/.local/lib/opencpn/*. There is no opencpn directory under ~/.local/lib/
3/ Start opencpn using --loglevel debug. Done
4/ Revert Active Setting to ubuntu-wx32-x86_64:22.04. Done
5/ Install a single plugin. Refreshed master catalogue and installed Logbook. No problem and opened for verification.
6/ Exit opencpn. Done
7/ Submit ~/.opencpn/opencpn.log. Attached as zipped file.

[opencpn.log.zip](https://github.com/jongough/testplugin_pi/files/11182917/opencpn.log.zip)

**Comment by LennartG-Sve** [comment]
@leamas , if I understand this right then the plugins generated based on TP v230 settings are carrying the wrong name, they are named 'ubuntu-x86_64:2204' and not 'ubuntu-wx32-x86_64:22.04'.
This explains why testplugin is not listed in the master catalogue and need be fixed. 
Nothing more to do until then.
Shall this issue remain open or shall I close it and open a new issue regarding the name error?

**Comment by leamas** [comment]
> if I understand this right then the plugins generated based on TP v230 settings are carrying the wrong name, they are named 'ubuntu-x86_64:2204' and not 'ubuntu-wx32-x86_64:22.04'.

Yes, besides that this is not really about the "name", it's about the XML metadata. However, Jon knows everything about this.

> Shall this issue remain open or shall I close it and open a new issue regarding the name error?

That I took part in this discussion was just to make sure that there was nothing wrong in core OpenCPN. As we can see, core OpenCPN does the right thing.

As you say, this is then about testplugin. I see no reason to file a new bug, this bug is filed against testplugin and we have just made sure that the problem  you observe is about testplugin, not OpenCPN. So,  basically this issue is now between you and @jongough,  I have not mush more to add.

That said, thanks for all your work!

**Comment by LennartG-Sve** [comment]
@leamas , Understand your point and comments and you are absolutely correct. I just wanted to make sure that this now is pointed towards TP and the FE2 base used. 
I now see two problems with the TP/FE2 configuration using v230:
1/ The naming 'problem' as discussed in detail above.
2/ The fact that both Windows versions fails with the same error message:
_C:\project\ocpn_project\build\Release\gpsodometer_pi.dll : fatal error LNK1120: 20 unresolved externals [C:\project\ocpn_project\build\gpsodometer_pi.vcxproj]
Command exited with code 1_

So: Thanks for your assistance in clearing the naming 'problem' out and I will await @jongough in due time.

**Comment by rgleason** [comment]
With regard to windows the windows error you report, I don't believe it is that simple, because I've found many plugins using TP 1.0.230.0 work.

**Comment by rgleason** [comment]
@leamas   Testplugin has been thoroughly tested for all OS and the [TP Template Table here](https://github.com/jongough/testplugin_pi/issues/284) is the result and was and was tested by Bdbcat too.  

Note the build list further down the thread:

- ubuntu-armhf-18.04-buster-armhf.xml <--- Has been added
- ubuntu-armhf-20.04-gtk3-focal-armhf.xml
- ubuntu-x86_64-18.04-bionic.xml
- ubuntu-x86_64-18.04-gtk3-bionic-gtk3.xml
- ubuntu-x86_64-20.04-gtk3-focal-gtk3.xml
- ubuntu-x86_64-22.04-jammy.xml

This has been a long standing decision made by you, not to use wx32 consistently.
As I understand it, you are now saying that all the plugins have to be rebuilt to add "-wx32" to the target for jammy.

When did this change?  Everyone must be informed of these changes.
Many of these plugins are not going to be rebuilt for 3 months.

**Comment by leamas** [comment]
> This has been a long standing decision made by you, not to use wx32 consistently.

This decision, as all decisions, is eventually Dave's.  That said, I have proposed it.

The usage is consistent. There is one simple rule: We add a suffix like -wx32 for a given platform version like 22.04 if and only if we build both wx3.0 and wx3.2 plugins for this platform. Since Jammy is the only platform which uses both wx3.0 and wx3.2 plugins, we use -wx32 here. See [above](https://github.com/jongough/testplugin_pi/issues/299#issuecomment-1500078564).

However, we don't really build ubuntu plugins, we build debian plugins. Basically, this boils down to that templates need to build `debian-<platform>`  using wx3.0 and `debian-wx32-<platform>`. If  the template just build these and avoids building plugins with an abi containing _ubuntu_ the pieces falls in place.

OTOH, as explained a plugin with a `_ubuntu-<platform>_`   ABI linked against wx3.2 i plain wrong. This is what seems to be happening here.

**Comment by rgleason** [comment]
OK, I agree it is not simple.  We had template TP 1.0.226 completely tested. In reviewing this, I think that perhaps Jon did not see these notes just below the[ TP Template Table](https://github.com/jongough/testplugin_pi/issues/284#issue-1551192013)

```
A. flatpak-x86_64-18.08-flatpak.xml <- No longer needed
A. raspbian-armhf-9.13-stretch-armhf.xml <- No longer needed
B. ubuntu-wx32-x86_64-22.04 <- Missing ADD Special case for jammy
C. ubuntu-x86_64-22.04-jammy.xml <- Remove, jammy must be wx32
D. ubuntu-x86_64-20.04-gtk3-focal-gtk3.xml<- Not needed, covered by debian-x86_64-11, when added
E. ubuntu-armhf-20.04-gtk3-focal-armhf.xml<- Not needed, covered by debian-armhf-11-bullseye-armhf.xml
F. ubuntu-armhf-18.04-buster-armhf.xml <- Not needed, covered by debian-armhf-10-buster-armhf.xml
G. ubuntu-x86_64-18.04-bionic.xml <- Not needed, covered by debian-x86_64-10-buster.xml
H. ubuntu-x86_64-18.04-gtk3-bionic-gtk3.xml<- Not needed, covered by debian-x86_64-10-buster.xml
```
In particular:

B. ubuntu-wx32-x86_64-22.04 <- Missing ADD Special case for jammy
C. ubuntu-x86_64-22.04-jammy.xml <- Remove, jammy must be wx32

**Comment by rgleason** [comment]
Very sorry Lennart.  Thank you for getting to the bottom of it.

**Comment by leamas** [comment]
That table is actually not correct. But let's wait until Jon is on board.

**Comment by leamas** [comment]
As I said, the easiest way to get it right is not build any ubuntu plugins at all, the debian plugins cover all ubuntu use cases.

**Comment by LennartG-Sve** [comment]
@rgleason , _Also your report states "2/ rm -f ~/.local/lib/opencpn/*. There is no opencpn directory under ~/.local/lib/"
...so where is your opencpn directory located?
or Did you indeed remove it from that location? (perhaps my misunderstanding)_
This is when I compile using git/appveyor. not locally so I'm using TP v230 configuration.

@leamas , there was a comment from Jon earlier that releases up to 5.6 used 'ubuntu-x86_64-22.04' so that version need then still be valid, right?

**Comment by rgleason** [comment]
@leamas   I have edited the actual [TP built files list here](https://github.com/jongough/testplugin_pi/issues/284#issuecomment-1484118255)  with comments afterwards referring to the Notes A-H directly after the [TP Template Table](https://github.com/jongough/testplugin_pi/issues/284#issue-1551192013).

Does this look correct to you now?
Also, if we remove all the Ubuntu metadata files, will users of OpenCPN 5.6.2 still have plugins that work?  If not, I think we should keep these ubuntu versions that don't fail, for awhile.

**Comment by leamas** [comment]
> there was a comment from Jon earlier that releases up to 5.6 used 'ubuntu-x86_64-22.04' so that version need then still be valid, right?


Yes, read my statement as "not build any ubuntu new plugins at all".  There is no need to rebuild 5.6.2/wx3.0 plugins, but plugin maintainers are of course free to do this if they want. In any case,  I'm advocating removal of existing, working plugins.

**Comment by rgleason** [comment]
Ok so, we are keeping 

`ubuntu-x86_64-22.04`

and all of the other ubuntu metadata and tarballs for OpenCPN 5.0 - 5.6.2 users. Correct?

What about this?  We don't have a armhf flatpak?   Which flatpak does it use?

`raspbian-armhf-9.13-stretch-armhf.xml <--- Remove, use flatpak`

**Comment by leamas** [comment]
> Ok so, we are keeping (I will change it back)
> ubuntu-x86_64-22.04

This will then be aimed for 5.6.2 using wx3.0. However, it is not really required, already 5.6.2 will try to load debian-x86_64:11 if it does not find any ubuntu match. So, why bother?

There is no armhf flatpak, so you need to build those.mn This is what shipdriver builds:
```
etadata/ShipDriver-3.2-android-arm64-A64-16.xml
metadata/ShipDriver-3.2-android-armhf-A32-16.xml
metadata/ShipDriver-3.2-darwin-wx32-10.xml
metadata/ShipDriver-3.2-debian-11.xml
metadata/ShipDriver-3.2-debian-12.xml
metadata/ShipDriver-3.2-debian-A32-11.xml
metadata/ShipDriver-3.2-debian-A32-12.xml
metadata/ShipDriver-3.2-debian-A64-11.xml
metadata/ShipDriver-3.2-debian-A64-12.xml
metadata/ShipDriver-3.2-debian-wx32-A64-11.xml
metadata/ShipDriver-3.2-debian-wx32.wx32-11.xml
metadata/ShipDriver-3.2-flatpak-22.08.xml
metadata/ShipDriver-3.2-flatpak-A64-22.08.xml
```

**Comment by bdbcat** [comment]
And this:
ubuntu-x86_64-wx32-22.04.xml <---0562 Must ADD for Jammy with wx32

Not needed.  O562 does not use wx32 on linux, ever.

**Comment by rgleason** [comment]
@leamas

> This will then be aimed for 5.6.2 using wx3.0. However, it is not really required, already 5.6.2 will try to load debian-x86_64:11 if it does not find any ubuntu match. So, why bother?

Does this apply to all of the older OpenCPN 5.6.2 Ubuntu builds?   - So we can safely remove them from the list and Users with 5.6.2 will be able to load the appropriate plugin?

`There is no armhf flatpak, so you need to build those. This is what shipdriver builds:`

So what are we missing exactly?  I don't see flatpak-armhf in Shipdriver builds.  Does this qualify as armhf?
flatpak-aarch64 | flatpak-aarch64-22.08 | 22.08 fltpk |   | wx32 gtk3 | Yes armhf?

**Comment by rgleason** [comment]
@bdbcat Bdbcat
> And this:
> ubuntu-x86_64-wx32-22.04.xml <---0562 Must ADD for Jammy with wx32
> Not needed. O562 does not use wx32 on linux, ever.

OK, what about Lennart's problem?  Don't we have to fix something?

**Comment by bdbcat** [comment]
@rgleason 
Lennart's problem on linux is well understood as a configuration problem on his system.
Problem on WIndows with TP/FE2 is not understood.  You are reporting that it works fine.  Is that correct?

**Comment by rgleason** [comment]
```
ShipDriver-3.2-android-arm64-A64-16.xml   - we have this
ShipDriver-3.2-android-armhf-A32-16.xml   - we have this
ShipDriver-3.2-darwin-wx32-10.xml    - we have this, but 11.4
ShipDriver-3.2-debian-11.xml    - we have this
ShipDriver-3.2-debian-12.xml     - we have this
ShipDriver-3.2-debian-A32-11.xml    --Is this x86_64?  if so, we have this
ShipDriver-3.2-debian-A32-12.xml    --Is this x86_64?   if so, we have this
ShipDriver-3.2-debian-A64-11.xml    --Is this arm64?     if so, we have this
ShipDriver-3.2-debian-A64-12.xml    --Is this arm64?     if so, we have this
ShipDriver-3.2-debian-wx32-A64-11.xml   --Is this arm64?  x86_64
ShipDriver-3.2-debian-wx32.wx32-11.xml             --what is this?
ShipDriver-3.2-flatpak-22.08.xml                          --what is this?
ShipDriver-3.2-flatpak-A64-22.08.xml   --Is this arm64?  if so, we have this.
```

We have these flatpaks, are they what we need?  --I am confused by the second one   x86_64 and arm64?
```
flatpak-aarch64-20.08-flatpak-arm64.xml
flatpak-x86_64-20.08-flatpak-arm64.xml
flatpak-x86_64-20.08-flatpak.xml
```

**Comment by rgleason** [comment]
```
Lennart's problem on linux is well understood as a configuration problem on his system.
Problem on WIndows with TP/FE2 is not understood. You are reporting that it works fine. Is that correct?
```
No, that is not what Alec concluded.

**Comment by bdbcat** [comment]
Rick...
Those are just file names.  It is the content of the XML that controls.
Probably the second one is non-sense, but the built XML should be examined to be sure.

**Comment by bdbcat** [comment]
No, that is not what Alec concluded.
How so?

**Comment by rgleason** [comment]
@bdbcat   https://github.com/jongough/testplugin_pi/issues/299#issuecomment-1500873894

**Comment by bdbcat** [comment]
Once again, OCPN core does not care about the name of the file.  It is the contents that count.
Further, there is no need to build ubuntu/22.04/wx32 plugins any more.  They are covered by debian build.
"Jon knows all this..."

**Comment by rgleason** [comment]
testplugin_pi-1.0.227.0-ubuntu-x86_64-22.04-jammy.xml

```
<target>ubuntu-x86_64</target>
<build-target>ubuntu</build-target>
<build-gtk>gtk3</build-gtk>
<target-version>22.04</target-version>
<target-arch>x86_64</target-arch>
<tarball-url>
```

Jon is away

**Comment by bdbcat** [comment]
testplugin_pi-1.0.227.0-ubuntu-x86_64-22.04-jammy.xml

```
<target>ubuntu-x86_64</target>
<build-target>ubuntu</build-target>
<build-gtk>gtk3</build-gtk>
<target-version>22.04</target-version>
<target-arch>x86_64</target-arch>
<tarball-url>

```
This one matches for O562 running on Jammy.  Not strictly necessary to build, builders discretion.

**Comment by bdbcat** [comment]
testplugin_pi-1.0.227.0-flatpak-aarch64-20.08-flatpak-arm64.xml

```
<target>flatpak-aarch64</target>
<build-target>flatpak</build-target>
<build-gtk></build-gtk>
<target-version>20.08</target-version>
<target-arch>aarch64</target-arch>
<tarball-url>

```
Bad name.
I agree.  Contents look fine for flatpak/arm64 build

**Comment by rgleason** [comment]
While out I thought to look up testplugin 1.0.226

```
<target>raspbian-armhf</target>
<build-target>raspbian</build-target>
<build-gtk>gtk2</build-gtk>
<target-version>9.13</target-version>
<target-arch>armhf</target-arch>
<tarball-url>
```

Ok. so I won't build it, and will remove it from all plugins in the master catalog.
Then the PIM will load what flatpak for OpenCPN 5.6.2 raspbian-armhf?    
We don't have a flatpak-armhf.

**Comment by bdbcat** [comment]
I see no reason to remove it, if it runs.  Some users may be running an ancient raspbian build on RPi.
There is Just no reason to build this one going forward.

**Comment by rgleason** [comment]
Dave see this [revised post ](https://github.com/jongough/testplugin_pi/issues/299#issuecomment-1500923153)please

**Comment by bdbcat** [comment]
Rick...
With correct understanding of host ABI settings and plugin targets, Lennart is now having no troubles on linux.
On Windows, his trouble is with local builds, probably due to incorrect wxWidgets version used for the build.
Summary:  There really are no problems with proposed Production OCPN from official builds, as I see it.
If you know otherwise, please share specifics.

**Comment by LennartG-Sve** [comment]
@bdbcat , Partly right, Linux is now sorted out as tp v230 did not have 'wx32' included as required for 5.7.1 builds, that will likely be fixed soon by Jon updating TP. I'll hold my horses here.
Windows: I do not do any local Windows builds (don't even have a Windows system), I'm talking about the build initiated thru appveyor when a version has been uploaded to git. Did a beta upload a few days ago and the error message is from that upload. 

Agree that this is not an OCPN problem, it is limited to TP-related builds.

**Comment by rgleason** [comment]
Thanks Dave and Alec.  We'll leave Testplugin at 1.0.230.

Lennart, I don't believe we have ever included wx32 in testplugin for linux or windows or any other OS, it is simply too big.  Only way is to install it yourself if you are going to build locally. Perhaps you would like to review the notes in the Dev Manual?   However I don't see anything about installing wxWidgets in "[Compiling Linux](https://opencpn-manuals.github.io/main/opencpn-dev/linux.html)"  perhaps Linux comes with wxWidgets?

**Comment by LennartG-Sve** [comment]
@rgleason , major misunderstanding! 
I've never asked for wxWidgets to be included in the plugin, I'm talking about the name of the two plugins required, 'ubuntu-wx32-x86_64-22.04' (wx32 version) for 5.7.1 and later respectively 'ubuntu-x86_64-22.04' (wx30 version) for versions up to 5.6.2. Today there is only one, a wx32-version, without the 'wx32' included in the name. I'm sure this will be taken care of by Jon in due time.

The remaining issue is instead the build initiated thru appveyor when the plugin was  uploaded to git, not compiled locally. I'm referencing the beta upload I did a few days ago and the error message from that upload. Both Windows versions failed with the same error message:

_C:\project\ocpn_project\build\Release\gpsodometer_pi.dll : fatal error LNK1120: 20 unresolved externals [C:\project\ocpn_project\build\gpsodometer_pi.vcxproj]
Command exited with code 1_

**Comment by leamas** [comment]
> Partly right, Linux is now sorted out as tp v230 did not have 'wx32' included as required for 5.7.1 builds, that will likely be fixed soon by Jon updating TP. 

The easier fix would be to just remove  this build completely. When OpenCPN does not find any ubuntu-* plugin it will try to load a debian-* plugin, in this case _debian-wx32-x86_64_ which will run just fine. This of course assuming that TP does build debian-wx32-x86_64. 

Again:  there is no need for any plugin abi like _ubuntu-wx32-*_,  all these use cases are covered by _debian-wx32-*_. Plain ubuntu-* linked to wx3.0 is not really useful either, 5.6.2 will also load debian plugins if applicable. 

And, bottom line: there is no need to remove existing, working  plugins for releases like Buster and  Bionic which basically are "frozen".

**Comment by leamas** [comment]
> I don't see anything about installing wxWidgets in "[Compiling Linux](https://opencpn-manuals.github.io/main/opencpn-dev/linux.html)" perhaps Linux comes with wxWidgets?

It's there. When you drag in the dependencies listed in the control file, this includes wxWidgets. 

After all, this is one of Linux major strengths: the dependencies are part of the OS. When they are not, for example wxsvg, we can add them to Linux rather than having to maintain it separately -- I have done this in  some cases, wxsvg is one of them.

**Comment by LennartG-Sve** [comment]
@leamas , was not aware that a debian alternative is searched for if 'the default ubuntu' is not found, sounds good. Then we may just drop some ubuntu variants to reduce the number of builds.

**Comment by LennartG-Sve** [comment]
@leamas , is this ubuntu -> debian autoseach/replacement also valid for 5.6.2 and earlier versions? Thinking about e.g. 'ubuntu-x86_64-22.04' being automagically replaced by 'debian-x86_64-22.04'?

**Comment by rgleason** [comment]
@LennartG-Sve 

> I've never asked for wxWidgets to be included in the plugin, I'm talking about the name of the two plugins required, 'ubuntu-wx32-x86_64-22.04' (wx32 version) for 5.7.1 and later respectively 'ubuntu-x86_64-22.04' (wx30 version) for versions up to 5.6.2. Today there is only one, a wx32-version, without the 'wx32' included in the name. I'm sure this will be taken care of by Jon in due time.

First of all it is not a wx32 version.  Lennart this change will never be made because for OpenCPN 5.7.1 we are using debian! 
 
We have in the builds 'ubuntu-x86_64-22.04' (wx30 version), 
**ubuntu-x86_64-22.04**   important metadata
<target>ubuntu-x86_64</target>     <--- There is no -wx32 here!!!
<build-target>ubuntu</build-target>
<build-gtk>gtk3</build-gtk>
<target-version>22.04</target-version>
<target-arch>x86_64</target-arch>

You can check the other metadata created by TP1.0.230 with this search https://cloudsmith.io/~opencpn/repos/testplugin-prod/packages/?q=1.0.230+metadata

Alec says that:
'ubuntu-x86_64-22.04' (wx30 version), if not found --> is _replaced_ by 'debian-x86_64-22.04' (Jammy)
I don't agree with that use of the language "replaced' as it leads to misunderstanding.  These new debian versions for Jammyt are using wx32 and gtk3 so they are only compatible with OpenCPN 5.7.1, and not 5.6.2!

They are only replaced by debian-x86_64-22.04' (Jammy) _after upgrade to OpenCPN 5.7.1_

I believe this is similar for the other ubuntu builds. So we are going to continue building them for awhile.

**Comment by rgleason** [comment]
PS: In our TP Template Table under **Ubuntu OS**  we have 
ubuntu-x86_64-22.04   as using wx30 and gtk3?    <---Is this correct?  I think perhaps not.

Are the other Ubuntu builds correctly listing wx and gtk?

**Comment by LennartG-Sve** [comment]
@rgleason , this is exactly in line with my two latest comments, ubuntu builds 5.7.1 upwards could well be replaced with debian builds, fine with me.

**Comment by rgleason** [comment]
@Lennart I am not in disagreement, I am trying to get the [table correct](https://github.com/jongough/testplugin_pi/issues/284) which I've changed as a result of this discussion.
1. ubuntu-x86_64-22.04 (Jammy) - Is this using wx30 and gtk3 as shown in the table?
2. Alec says if ubuntu-x86_64-22.04 (Jammy) is not found that debian-x86_64:11  is loaded.
3. Do the wx30 and gtk under _Ubuntu OS_ look right in the table?
4. Could we indeed remove all the Ubuntu builds and would OpenCPN 5.6.2 Users still be able to load the plugins using the TP 1.0.230 version only?

**Comment by bdbcat** [comment]
Corrections:
B.  Still required in the catalog for O562.  However, there is no pressing need to rebuild it.  Consider it "frozen".  Raspbian Stretch is very much obsolete, few users still on it.

Final note (I. ?)  ubuntu-wx32-x86_64-22.04 (Jammy) is **absolutely required for O571**.  It's counterpart for O562 will be covered by "debian-x86_64-11"

tl;dr;
The only "ubuntu" build required going forward is "ubuntu-wx32-x86_64-22.04 (Jammy)".  This is a special case, unavoidable.
As you will note, the build matrix is significantly simplified now.
And one more point:  The templates may currently build all sorts of "ubuntu" flavors.  It is up to you, the packager, to decide how to update the catalog.

**Comment by LennartG-Sve** [comment]
@rgleason , Have limited knowledge apart from the ubuntu 22.04 that I've been working with:
The generated plugin 'ubuntu-x86_64-22.04' (number 1 in your list) is using wx32 and gtk3. That is the reason the TP v230 generated plugins does not show up in PIM.
The naming convention states that this name is for wz30 plugins (pre 5.7.1), else 'wx32' should be included in the name.  The currently generated plugin 'ubuntu-x86_64-22.04'  is indeed ubuntu-based.

**Comment by rgleason** [comment]
@bdbcat   is this statement true?

> 'ubuntu-x86_64-22.04' (number 1 in your list) is using wx32 and gtk3. That is likely (?) the reason the TP v230 generated plugins does not show up in PIM.

ubuntu-x86_64-22.04 was tested by Ozolli and it does show up in PIM and does work.  What is Lennart's problem?

**Comment by bdbcat** [comment]
Lennart....
We need more info.
1.  Exactly where/how did you load (or build?) OCPN 571 for your ubuntu 22.04?
2. Please show again the OCPN logfile from a clean startup.
Thanks
Dave

**Comment by bdbcat** [comment]
Rick:
_Also I recall something about Jammy always being wx32, therefore it is not explictly called for in the <target> Is that true?
-a case of conventions being broken, to move forward?_

Not true.
Jammy itself is wx-agnostic.  It will use wx30 if O562 is installed, and wx32 if O571 is installed.  Thus, we need the "wx32" in the pluigin target XML to distinguish which plugins to load.

However, all supported ubuntu LTS after Jammy (i.e. 24.04 due in one year) will be able to drop the "wx32" from the plugin target.

**Comment by bdbcat** [comment]
Rick...
I think Lennart's problem is this:
In master catalog, looking at TestPlugin  v1.0.230.0, there is no metadata entry for wx32 on jammy.  There is one entry that is close, and would work for O562.   But not for O571

```
  <plugin version="1">
    <name> Testplugin </name>
    <version> 1.0.230.0 </version>
    <release> 0 </release>
    <summary> Plugin to test examples of the ODAPI and JSON interface for ODRAW </summary>
    <api-version> 1.16 </api-version>
    <open-source> yes </open-source>
    <author> Jon Gough </author>
    <source> https://github.com/jongough/testplugin_pi </source>
    <description>
testplugin Plugin is used to test out the ODraw API and demonstrate how to use it successfully from another plugin
</description>
    <target>ubuntu-x86_64</target>
    <build-target>ubuntu</build-target>
    <build-gtk>gtk3</build-gtk>
    <target-version>22.04</target-version>
    <target-arch>x86_64</target-arch>
    <tarball-url>
https://dl.cloudsmith.io/public/opencpn/testplugin-prod/raw/names/testplugin_pi-1.0.230.0-ubuntu-x86_64-22.04-jammy-tarball/versions/v1.0.230.0/testplugin_pi-1.0.230.0-ubuntu-x86_64-22.04-jammy.tar.gz
</tarball-url>
    <info-url> https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:opencpn_user_manual:plugins:safety:odraw1.6_pi </info-url>
  </plugin>

```

**Comment by rgleason** [comment]
Yes, this needs to be added to the builds and I have so noted this in [TP Template Table](https://github.com/jongough/testplugin_pi/issues/284#issuecomment-1484118255) at the bottom of Debian, noted as "special"

> In master catalog, looking at TestPlugin v1.0.230.0, there is no metadata entry for wx32 on jammy. There is one entry that is close, and would work for O562. But not for O571

**Comment by rgleason** [comment]
> The only "ubuntu" build required going forward is "ubuntu-wx32-x86_64-22.04 (Jammy)"., and ARMxx variants. This is a special case, unavoidable.

So we need to add "special" builds for Jammy  ?

- ubuntu-armhf-wx32-22.04
- ubuntu-arm64-wx32-22.04
- ubuntu-x86_64-wx32-22.04

**Comment by bdbcat** [comment]
So, we now can agree that Lennart's problem with O571 on ubuntu 22.04  is understood?

So we need to add "special" builds for Jammy ?

    ubuntu-armhf-wx32-22.04
    ubuntu-arm64-wx32-22.04
    ubuntu-x86_64-wx32-22.04
Yes, exactly.

**Comment by rgleason** [comment]
@bdbcat 
in[ TP Template Table](https://github.com/jongough/testplugin_pi/issues/284#issuecomment-1484118255)
Are debian-armhf-11,  debian-arm64-10, debian-arm64-11, debian-x86_64-11  all  "O562 Frozen" (in the last column?)

**Comment by rgleason** [comment]
Also we do not appear to have:

debian-wx32-x86_64 | debian-wx32-x86_64-11 | 11 bullseye | Jammy * | wx32 gtk3 | ADD ppa?
-- | -- | -- | -- | -- | --
debian-wx32-armhf | debian-wx32-armhf-11 | 11 bullseye | Jammy * | wx32 gtk3 | ADD ppa?
debian-wx32-arm64 | debian-wx32-arm64-11 | 11 bullseye | Jammy * | wx32 gtk3 | ADD ppa?

**Comment by leamas** [comment]
> So we need to add "special" builds for Jammy ?

No.

The following applies to Focal (20.04) and Jammy (22.04).  For older releases the simple rule is that while there is no need to build new plugin versions, we keep the existing entries in the catalog as-is. In other words, versions before Focal (and Debian Bulllseye) are "frozen".

For Ubuntu Focal, Ubuntu Jammy, Debian bullseye and Raspbian/11 we build two  kinds of plugins: `debian-<platform>:11` and `debian-wx32-<platform>`.  This is basically the end of the roadmap described in https://github.com/OpenCPN/OpenCPN/discussions/2502.

There is no need to build ubuntu plugins since 5.6.2 because:
  - Since 5.6.2 opencpn versions linked against wx3.0 will fall back to using `debian-<platform>:11` if not finding a ubuntu-based plugin
  - Likewise, since 5.6.2 opencpn versions linked against wx3.2 will fall back to using `debian-wx32-<platform>:11` if not finding a ubuntu-based plugin.

The whole idea behind the discussions in https://github.com/OpenCPN/OpenCPN/discussions/2502 is that we should build debian plugins rather  then the ubuntu counterparts since these plugins runs in a well defined way in all the debian derivatives we care about: Ubuntu and Raspbian. This means that while building ubuntu plugins is possible, it's basically wasted work. 

> The only "ubuntu" build required going forward is "ubuntu-wx32-x86_64-22.04 (Jammy)"., and ARMxx variants. This is a special case, unavoidable.

What am I missing here? why wouldn't `debian-wx32-x86:64:11` work? "Mystified"

**Comment by rgleason** [comment]
With regard to flatpak builds,  in the Table we show:

flatpak-x86_64 | flatpak-x86_64-20.08 | 20.08 fltpk |   | wx30 gtk3 | O56.2 Frozen
-- | -- | -- | -- | -- | --
flatpak-aarch64 | flatpak-aarch64-20.08 | 20.08 fltpk |   | wx30 gtk3 | O56.2 Frozen
flatpak-x86_64 | flatpak-x86_64-22.08 | 22.08 fltpk |   | wx32 gtk3 | Yes
flatpak-arm64 | flatpak-arm64-22.08 | 22.08 fltpk |   | wx32 gtk3 |  
flatpak-aarch64 | flatpak-aarch64-22.08 | 22.08 fltpk |   | wx32 gtk3 | 

and the metadata build list shows:   (I don't think this is correct... 
...see what is in a [cloudsmith search in testplugin-prod   "1.0.230 metadata flatpak"](https://cloudsmith.io/~opencpn/repos/testplugin-prod/packages/?q=1.0.230+metadata+flatpak)

- flatpak-aarch64-20.08-flatpak-arm64.xml
- flatpak-x86_64-20.08-flatpak-arm64.xml
- flatpak-x86_64-20.08-flatpak.xml

Metadata files from [cloudsmith search in testplugin-prod   "1.0.230 metadata flatpak"](https://cloudsmith.io/~opencpn/repos/testplugin-prod/packages/?q=1.0.230+metadata+flatpak) contain:

**testplugin_pi-1.0.230.0-flatpak-aarch64-20.08-flatpak-arm64.xml**
```
<target>flatpak-aarch64</target>
<build-target>flatpak</build-target>
<build-gtk></build-gtk>
<target-version>20.08</target-version>
<target-arch>aarch64</target-arch>
```
**testplugin_pi-1.0.230.0-flatpak-x86_64-20.08-flatpak.xml**
```
<target>flatpak-x86_64</target>
<build-target>flatpak</build-target>
<build-gtk></build-gtk>
<target-version>20.08</target-version>
<target-arch>x86_64</target-arch>
```


**testplugin_pi-1.0.230.0-flatpak-x86_64-22.08-flatpak.xml**    <<---This seems to be doubled.
```
<target>flatpak-x86_64</target>
<build-target>flatpak</build-target>
<build-gtk></build-gtk>
<target-version>22.08</target-version>
<target-arch>x86_64</target-arch>
```

So it appears we are missing quite a lot of flatpak in this version 1.0.230

**Comment by rgleason** [comment]
@leamas 

>     So we need to add "special" builds for Jammy ?

> No.

Yes this is true for the shipdriver approach, but not for the way Jon is doing TP

**Comment by leamas** [comment]
> So we need to add "special" builds for Jammy ?

No.

The following applies to Focal (20.04) and Jammy (22.04).  For older releases the simple rule is that while there is no need to build new plugin versions, we keep the existing entries in the catalog as-is. In other words, versions before Focal (and Debian Bulllseye) are "frozen".

For Ubuntu Focal, Ubuntu Jammy, Debian bullseye and Raspbian/11 we build two  kinds of plugins: `debian-<platform>:11` and `debian-wx32-<platform>`.  This is basically the end of the roadmap described in https://github.com/OpenCPN/OpenCPN/discussions/2502.

There is no need to build ubuntu plugins since 5.6.2 because:
  - Since 5.6.2 opencpn versions linked against wx3.0 will fall back to using `debian-<platform>:11` if not finding a ubuntu-based plugin
  - Likewise, since 5.6.2 opencpn versions linked against wx3.2 will fall back to using `debian-wx32-<platform>:11` if not finding a ubuntu-based plugin.

The whole idea behind the discussions in https://github.com/OpenCPN/OpenCPN/discussions/2502 is that we should build debian plugins rather  then the ubuntu counterparts since these plugins runs in a well defined way in all the debian derivatives we care about: Ubuntu and Raspbian. This means that while building ubuntu plugins is possible, it's basically wasted work. 

> The only "ubuntu" build required going forward is "ubuntu-wx32-x86_64-22.04 (Jammy)"., and ARMxx variants. This is a special case, unavoidable.

What am I missing here? why wouldn't `debian-wx32-x86:64:11` work? "Mystified"

**Comment by rgleason** [comment]
Thanks Alec, I am just following here and trying to get this table accurate, correct and right.  I will wait for bdbcat...

**Comment by bdbcat** [comment]
Alec...
What am I missing here? why wouldn't debian-wx32-x86_64:11 work? "Mystified"
How can one build a wx32 plugin (or, indeed, OCPN core) for debian 11?  IIUC, debian 11 is still offering wx30 in it's default repo.
https://packages.debian.org/bullseye/wx-common

**Comment by bdbcat** [comment]
Alec...
We do not build OCPN nor plugins against wx32 for focal.  The cutover is/was jammy, and this as a special case because jammy is hydrid, looks like debain 12 in some ways, but offers only wx30 in default repo.

**Comment by rgleason** [comment]
You guys are marvels. Thank you.  Have a Happy Easter!  off to some friends.

**Comment by leamas** [comment]
> How can one build a wx32 plugin (or, indeed, OCPN core) for debian 11? IIUC, debian 11 is still offering wx30 in it's default repo.

... which means that we have to use our  own w32 backport to Bullseye. This is the same problem as building the wx32 version on Jammy, solved there. Shipdriver has also solved this.

**Comment by leamas** [comment]
>  We do not build OCPN nor plugins against wx32 for focal. The cutover is/was jammy,

Agreed. Still, this is no reason to not stick to the simplest form: Use debian everywhere.

**Comment by leamas** [comment]
>  We do not build OCPN nor plugins against wx32 for focal. The cutover is/was jammy,

Agreed. Still, this is no reason to not stick to the simplest form: Use debian everywhere.

**Comment by bdbcat** [comment]
Hmmm...  I will verify that with o-charts.
SB

**Comment by bdbcat** [comment]
Yes, OK, that is what I am doing, sure.
I agree, use debian wherever possible on new builds.
Sorry for the additional FUD-factor.

**Comment by rgleason** [comment]
So in lieu of the ubuntu arm64, armhf, x86_64 wx32 builds, do we need to build these?

debian-wx32-x86_64 | debian-wx32-x86_64-11 | 11 bullseye | Jammy * | wx32 gtk3 | ADD ppa?
-- | -- | -- | -- | -- | --
debian-wx32-armhf | debian-wx32-armhf-11 | 11 bullseye | Jammy * | wx32 gtk3 | ADD ppa?
debian-wx32-arm64 | debian-wx32-arm64-11 | 11 bullseye | Jammy * | wx32 gtk3 | ADD ppa?
debian-x86_64 | debian-x86_64-12 | 12 bookworm | Jammy * | wx32 gtk3 | ADD
debian-arm64 | debian-arm64-12 | 12 bookworm | Jammy * | wx32 gtk3 | ADD
debian-armhf | debian-armhf-12 | 12 bookworm | Jammy * | wx32  gtk3 | ADD

**Comment by bdbcat** [comment]
Rick...
I have edited the build list, here.  Imposed a little order to the entries.  Added required debian-wx32 builds
About the  debian wx30 builds:  They need to be kept in the build for cases where bullseye and buster are running OCPN571 wxthout wx32.
```

android-arm64-16-android-arm64.xml
android-armhf-16-android-armhf.xml

darwin-wx32-x86_64-11.4-macos.xml
darwin-x86_64-11.4-macos.xml

flatpak-aarch64-20.08-flatpak-arm64.xml
flatpak-x86_64-20.08-flatpak-arm64.xml
flatpak-x86_64-20.08-flatpak.xml

msvc-x86_64-10.0.14393-MSVC.xml
msvc-x86_64-wx32-10.0.17763-MSVC.xml

debian-armhf-10-buster-armhf.xml
debian-arm64-10-buster-arm64.xml
debian-armhf-10-gtk3-buster-armhf.xml
debian-x86_64-10-buster.xml

debian-arm64-11-bullseye-arm64.xml
debian-armhf-11-bullseye-armhf.xml
debian-x86_64-11-bullseye.xml

raspbian-armhf-9.13-stretch-armhf.xml <--- O562 FROZEN

ubuntu-armhf-18.04-buster-armhf.xml <--- O562 Remove, use debian-armhf-10-buster
ubuntu-armhf-20.04-gtk3-focal-armhf.xml <---O562 Remove, use debian-armhf-11-bullseye
ubuntu-x86_64-18.04-bionic.xml <---O562 Remove, use debian-x86_64-10-buster
ubuntu-x86_64-18.04-gtk3-bionic-gtk3.xml <---O562 Remove, use debian-x86_64-10-buster
ubuntu-x86_64-20.04-gtk3-focal-gtk3.xml <--O562 Remove, use debian-x86_64-11-bullseye
ubuntu-x86_64-22.04-jammy.xml <---O562 Remove uses debian-x86_64-11-bullseye

Add:  Trying to follow your name configuration

debian-wx32-arm64-10-buster-arm64.xml
debian-wx32-armhf-10-gtk3-buster-armhf.xml
debian-wx32-armhf-10-buster-armhf.xml
debian-wx32-x86_64-10-buster.xml

debian-wx32-arm64-11-bullseye-arm64.xml
debian-wx32-armhf-11-bullseye-armhf.xml
debian-wx32-x86_64-11-bullseye.xml

```
If possible, will you keep the re-arrangement intact.  Make it easier to categorize the work.

**Comment by rgleason** [comment]
Great. Thanks very much. Happy Easter.

**Comment by LennartG-Sve** [comment]
@bdbcat , 
1/  I'm using the PPA version, version is 'OpenCPN 5.7.1+7936+fccc1422c-ubuntu22-04.1', a start / stop log is attached.
[ocpn_clean.log.zip](https://github.com/jongough/testplugin_pi/files/11188368/ocpn_clean.log.zip)

2/ _I think Lennart's problem is this:
In master catalog, looking at TestPlugin v1.0.230.0, there is no metadata entry for wx32 on jammy. There is one entry that is close, and would work for O562. But not for O571_
I fully agree.

**Comment by LennartG-Sve** [comment]
Guys, now that the 'main issue' seems sorted out I'd like to turn the attention towards the second issue i had when building beta plugins uploading to git:
The build process initiated thru AppVeyor, referencing the beta upload I did a few days ago, and the error message from that build. Both Windows versions failed with the same error message:

C:\project\ocpn_project\build\Release\gpsodometer_pi.dll : fatal error LNK1120: 20 unresolved externals [C:\project\ocpn_project\build\gpsodometer_pi.vcxproj]
Command exited with code 1

**Comment by leamas** [comment]
> Guys, now that the 'main issue' seems sorted out I'd like to turn the attention towards the second issue i had when building beta plugins uploading to git:

Please don't do this, file  a new issue instead.

**Comment by LennartG-Sve** [comment]
@leamas , OK. It was originally a part of this issue but you got a point.This discussion has got to extensive so I'm ready to close it in a day or two if no-one objects.

**Comment by leamas** [comment]
There are multiple errors in that table. If I was to use such a table I would split it  in two parts. One would list the "frozen" builds we still keep but doesn't update. The other would list the builds we actually update, this is a much smaller list.

IMHO it's not really possible to review the current state. Just some drive-by remarks:
  - There will never be any need for a buster-wx32 plugin
  - Jammy used debian:11, not debian:12

**Comment by leamas** [comment]
Again, here is the shipdriver builds as a reference. Naming legend:
  - no platform means x86_64/amd64
  - A64 means arm64/aarch64
  - A32 means armhf

```
ShipDriver-3.2-android-arm64-A64-16.xml
ShipDriver-3.2-android-armhf-A32-16.xml
ShipDriver-3.2-darwin-wx32-10.xml
ShipDriver-3.2-debian-11.xml
ShipDriver-3.2-debian-12.xml
ShipDriver-3.2-debian-A32-11.xml
ShipDriver-3.2-debian-A32-12.xml
ShipDriver-3.2-debian-A64-11.xml
ShipDriver-3.2-debian-A64-12.xml
ShipDriver-3.2-debian-wx32-A64-11.xml
ShipDriver-3.2-debian-wx32.wx32-11.xml
ShipDriver-3.2-flatpak-22.08.xml
ShipDriver-3.2-flatpak-A64-22.08.xml
```

**Comment by rgleason** [comment]
Thanks Alec for your thoughtful drive by comments, and take them into consideration. I will wait for Bdbcat to advise. This is not a process that I enjoy, and you make it so much easier.

**Comment by LennartG-Sve** [comment]
This is apparently sorted out, no comments for 3 days, closing.

**Comment by rgleason** [comment]
Lennart, dispite the pain you went through, I think this was a big help to focus on the changes/improvements needed for Testplugin builds, and we now have a good roadmap.

Next step is to try to get the new debian builds done, and I am in the process of reviewing how Jon accomplishes that (most of the additional debian builds will simply need changes to the ci/config.yml I think).
